File: julia/strassen_matmul.jl
using Base.Threads

include("iterative_matmul.jl")

"""
    next_power_of_2(n::Int)

Find the next power of 2 greater than or equal to n.
"""
function next_power_of_2(n::Int)
  n <= 0 && return 1
  power = 1
  while power < n
    power *= 2
  end
  return power
end

"""
    strassen_matmul(A::AbstractMatrix{T}, B::AbstractMatrix{T}; threshold::Int=32) where T

Strassen's matrix multiplication algorithm with optional parallelization.
Recursively divides matrices into quadrants and computes 7 products instead of 8,
reducing complexity from O(n³) to O(n^2.807).

Arguments:
- A, B: Input matrices to multiply
- threshold: Minimum size to switch to standard multiplication (default: 32)
"""
function strassen_matmul(A::AbstractMatrix{T}, B::AbstractMatrix{T}; threshold::Int=32) where T
  m, n = size(A)
  q, p = size(B)

  if n != q
    throw(DimensionMismatch("Matrix dimensions must agree: A is $(m)x$(n), B is $(q)x$(p)"))
  end

  if Threads.nthreads() <= 1
    throw(ArgumentError("strassen_matmul requires multiple threads to run"))
  end

  # Pad matrices to the nearest power of 2 for optimal Strassen algorithm
  padded_m = next_power_of_2(m)
  padded_n = next_power_of_2(n)
  padded_p = next_power_of_2(p)

  # Pad if necessary
  if m != padded_m || n != padded_n
    A_padded = zeros(T, padded_m, padded_n)
    A_padded[1:m, 1:n] = A
    A = A_padded
  end

  if n != padded_n || p != padded_p
    B_padded = zeros(T, padded_n, padded_p)
    B_padded[1:q, 1:p] = B
    B = B_padded
  end

  result = _strassen_recursive(A, B, threshold)

  # Extract the original size result
  return result[1:m, 1:p]
end

"""
    _strassen_recursive(A, B, threshold)

Internal recursive function for Strassen multiplication.
"""
function _strassen_recursive(A::AbstractMatrix{T}, B::AbstractMatrix{T}, threshold::Int) where T
  m, n = size(A)
  q, p = size(B)

  # Base case: use iterative multiplication for small matrices
  if m <= threshold || n <= threshold || p <= threshold
    return iterative_matmul(A, B)
  end

  # Calculate split points (matrices are already padded to power of 2)
  m_half = m ÷ 2
  n_half = n ÷ 2
  p_half = p ÷ 2

  # Divide A into quadrants
  A11 = @view A[1:m_half, 1:n_half]
  A12 = @view A[1:m_half, n_half+1:n]
  A21 = @view A[m_half+1:m, 1:n_half]
  A22 = @view A[m_half+1:m, n_half+1:n]

  # Divide B into quadrants
  B11 = @view B[1:n_half, 1:p_half]
  B12 = @view B[1:n_half, p_half+1:p]
  B21 = @view B[n_half+1:n, 1:p_half]
  B22 = @view B[n_half+1:n, p_half+1:p]

  # Strassen's algorithm: compute 7 products instead of 8
  # M1 = (A11 + A22) * (B11 + B22)
  # M2 = (A21 + A22) * B11
  # M3 = A11 * (B12 - B22)
  # M4 = A22 * (B21 - B11)
  # M5 = (A11 + A12) * B22
  # M6 = (A21 - A11) * (B11 + B12)
  # M7 = (A12 - A22) * (B21 + B22)
  #
  # C11 = M1 + M4 - M5 + M7
  # C12 = M3 + M5
  # C21 = M2 + M4
  # C22 = M1 - M2 + M3 + M6

  # Parallel computation using threads
  # Spawn 7 tasks for the 7 Strassen products
  t1 = Threads.@spawn _strassen_recursive(A11 + A22, B11 + B22, threshold)
  t2 = Threads.@spawn _strassen_recursive(A21 + A22, B11, threshold)
  t3 = Threads.@spawn _strassen_recursive(A11, B12 - B22, threshold)
  t4 = Threads.@spawn _strassen_recursive(A22, B21 - B11, threshold)
  t5 = Threads.@spawn _strassen_recursive(A11 + A12, B22, threshold)
  t6 = Threads.@spawn _strassen_recursive(A21 - A11, B11 + B12, threshold)
  t7 = Threads.@spawn _strassen_recursive(A12 - A22, B21 + B22, threshold)

  # Fetch results
  M1 = fetch(t1)
  M2 = fetch(t2)
  M3 = fetch(t3)
  M4 = fetch(t4)
  M5 = fetch(t5)
  M6 = fetch(t6)
  M7 = fetch(t7)

  # Combine results to form C
  C11 = M1 + M4 - M5 + M7
  C12 = M3 + M5
  C21 = M2 + M4
  C22 = M1 - M2 + M3 + M6

  # Combine quadrants
  C = [C11 C12; C21 C22]

  return C
end

File: rust/src/strassen_matmul.rs
use crate::iterative_matmul::iterative_matmul;
use crate::matrix::Matrix;
use rayon::prelude::*;

/// Strassen matrix multiplication algorithm with parallelization.
/// Uses 7 recursive multiplications instead of 8 by computing intermediate
/// products and combining them cleverly.
///
/// # Arguments
/// * `a` - First input matrix
/// * `b` - Second input matrix
/// * `threshold` - Minimum size to switch to standard multiplication
///
/// # Returns
/// Result matrix
pub fn strassen_matmul(a: &Matrix, b: &Matrix, threshold: usize) -> Matrix {
	let m = a.rows;
	let n = a.cols;
	let q = b.rows;
	let p = b.cols;

	if n != q {
		panic!(
			"Matrix dimensions must agree: A is {}x{}, B is {}x{}",
			m, n, q, p
		);
	}

	if rayon::current_num_threads() <= 1 {
		panic!("strassen_matmul requires multiple threads to run");
	}

	strassen_recursive(a, b, threshold)
}

/// Internal recursive function for Strassen multiplication.
fn strassen_recursive(a: &Matrix, b: &Matrix, threshold: usize) -> Matrix {
	let m = a.rows;
	let n = a.cols;
	let p = b.cols;

	// Base case: use iterative implementation for fair comparison
	if m <= threshold || n <= threshold || p <= threshold {
		return iterative_matmul(a, b);
	}

	// Divide matrices into quadrants
	let m_half = m / 2;
	let n_half = n / 2;
	let p_half = p / 2;

	// Divide A into quadrants
	let a11 = a.submatrix(0, m_half, 0, n_half);
	let a12 = a.submatrix(0, m_half, n_half, n);
	let a21 = a.submatrix(m_half, m, 0, n_half);
	let a22 = a.submatrix(m_half, m, n_half, n);

	// Divide B into quadrants
	let b11 = b.submatrix(0, n_half, 0, p_half);
	let b12 = b.submatrix(0, n_half, p_half, p);
	let b21 = b.submatrix(n_half, n, 0, p_half);
	let b22 = b.submatrix(n_half, n, p_half, p);

	// Compute the 7 Strassen products
	// M1 = (A11 + A22) * (B11 + B22)
	// M2 = (A21 + A22) * B11
	// M3 = A11 * (B12 - B22)
	// M4 = A22 * (B21 - B11)
	// M5 = (A11 + A12) * B22
	// M6 = (A21 - A11) * (B11 + B12)
	// M7 = (A12 - A22) * (B21 + B22)

	let mut a11_a22 = a11.clone();
	a11_a22.add(&a22);

	let mut b11_b22 = b11.clone();
	b11_b22.add(&b22);

	let mut a21_a22 = a21.clone();
	a21_a22.add(&a22);

	let mut b12_b22 = b12.clone();
	b12_b22.sub(&b22);

	let mut b21_b11 = b21.clone();
	b21_b11.sub(&b11);

	let mut a11_a12 = a11.clone();
	a11_a12.add(&a12);

	let mut a21_a11 = a21.clone();
	a21_a11.sub(&a11);

	let mut b11_b12 = b11.clone();
	b11_b12.add(&b12);

	let mut a12_a22 = a12.clone();
	a12_a22.sub(&a22);

	let mut b21_b22 = b21.clone();
	b21_b22.add(&b22);

	// Prepare all inputs for parallel computation
	let inputs: Vec<(&Matrix, &Matrix)> = vec![
		(&a11_a22, &b11_b22), // M1
		(&a21_a22, &b11),     // M2
		(&a11, &b12_b22),     // M3
		(&a22, &b21_b11),     // M4
		(&a11_a12, &b22),     // M5
		(&a21_a11, &b11_b12), // M6
		(&a12_a22, &b21_b22), // M7
	];

	// Parallel computation using rayon
	let results: Vec<Matrix> = inputs
		.into_par_iter()
		.map(|(a_sub, b_sub)| strassen_recursive(&a_sub, &b_sub, threshold))
		.collect();

	let [m1, m2, m3, mut m4, mut m5, mut m6, mut m7]: [Matrix; 7] = results.try_into().unwrap();

	// Combine the products to get result quadrants
	// C11 = M1 + M4 - M5 + M7
	// C12 = M3 + M5
	// C21 = M2 + M4
	// C22 = M1 - M2 + M3 + M6

	m7.add(&m1).add(&m4).sub(&m5);
	let c11 = m7;

	m5.add(&m3);
	let c12 = m5;

	m4.add(&m2);
	let c21 = m4;

	m6.add(&m1).sub(&m2).add(&m3);
	let c22 = m6;

	// Combine quadrants
	Matrix::combine_quadrants(&c11, &c12, &c21, &c22)
}
